eclass/kernel-build.eclass

```diff
@@ -369,6 +416,87 @@ kernel-build_src_install() {
 		secureboot_sign_efi_file "${ED}${kernel_dir}/${image_path}"
 	fi
 
+	if [[ ${KERNEL_IUSE_GENERIC_UKI} ]]; then
+		if use generic-uki; then
+			local image="${ED}${kernel_dir}/${image_path}"
+
+			# NB: if you pass a path that does not exist or is not a regular
+			# file/directory, dracut will silently ignore it and use the default
+			# https://github.com/dracutdevs/dracut/issues/1136
+			> "${T}"/empty-file || die
+			mkdir -p "${T}"/empty-directory || die
+
+			local dracut_args=(
+				--conf "${T}/empty-file"
+				--confdir "${T}/empty-directory"
+				--kernel-image "${image}"
+				--kmoddir "${ED}/lib/modules/${PV}-gentoo-dist"
+				--verbose
+				--no-hostonly
+				--no-hostonly-cmdline
+				--no-hostonly-i18n
+				--no-machineid
+				--nostrip
+				--no-uefi
+				--early-microcode
+				--reproducible
+				--ro-mnt
+				--modules="base bash btrfs cifs crypt crypt-gpg crypt-loop dbus dbus-daemon dm dmraid drm dracut-systemd fido2 i18n iscsi fs-lib kernel-modules kernel-network-modules kernel-modules-extra lunmask lvm nbd mdraid modsign network network-manager nfs nvdimm nvmf pcsc pkcs11 plymouth qemu qemu-net resume rngd rootfs-block shutdown systemd systemd-ac-power systemd-ask-password systemd-initrd systemd-integritysetup systemd-pcrphase systemd-sysusers systemd-udevd systemd-veritysetup terminfo tpm2-tss udev-rules uefi-lib usrmount virtiofs"
+			)
+
+			# Tries to update ld cache
+			addpredict /etc/ld.so.cache~
+			dracut "${dracut_args[@]}" "${image%/*}/initrd" || die "Failed to generate initramfs"
+
+			local ukify_args=(
+				--linux="${image}"
+				--initrd="${image%/*}/initrd"
+				--cmdline="root=/dev/gpt-auto-root ro quiet splash"
+				--uname="${PV}"
+				--output="${image%/*}/uki.efi"
+			)
+
+			if [[ ${KERNEL_IUSE_SECUREBOOT} ]]; then
+				if use secureboot; then
+					ukify_args+=(
+						--signtool=sbsign
+						--secureboot-private-key="${SECUREBOOT_SIGN_KEY}"
+						--secureboot-certificate="${SECUREBOOT_SIGN_CERT}"
+					)
+					if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
+						ukify_args+=(
+							--signing-engine="pkcs11"
+						)
+					else
+						# Sytemd-measure does not currently support pkcs11
+						ukify_args+=(
+							--measure
+							--pcrpkey="${ED}/usr/src/linux-${PV}-gentoo-dist/certs/signing_key.x509"
+							--pcr-private-key="${SECUREBOOT_SIGN_KEY}"
+							--phases="enter-initrd"
+							--pcr-private-key="${SECUREBOOT_SIGN_KEY}"
+							--phases="enter-initrd:leave-initrd enter-initrd:leave-initrd:sysinit enter-initrd:leave-initrd:sysinit:ready"
+						)
+					fi
+				fi
+			fi
+
+			# Find ukify on both split-usr/merged-usr, set fallback
+			local ukify=ukify
+			for path in "${BROOT}/usr/lib/systemd/ukify" "${BROOT}/lib/systemd/ukify"; do
```

Comment:
Can we ask pkg-config for help?

Reply:
> Can we ask pkg-config for help?

Maybe, I'll give it a try.

Reply:
It's fine if you do it as a followup given the paths are pretty reliable.

Reply:
An additional complication is that systemd-255 and up install `(/usr)/lib/systemd/ukify` as backwards compatibility symlink to `/usr/bin/ukify`. I don't know if this is planned to be removed at some point, but if it is then things would break here if I rely here on `pkg-config` telling me if systemd installs stuff into `/usr/lib/systemd` or `/lib/systemd`.

This is why in the current code I have the `ukify=ukify` fallback which ensures that things will continue to work if systemd removes this backwards compatibility symlink in the future.